### PR TITLE
[Feature] Instructor dashboards with second correction

### PR DESCRIPTION
**Pull request by SimonEntholzer**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
With the addition of the second correction feature we want that instructors are able to see each individual assessment. Those should be accessible over the instructor dashboard of the specific exercise type.

### Description
Instructors now can access and conduct individual assessments for each correction round for modeling and text exercises.

### Steps for Testing

**_Edit_: Please double check that all submissions are shown. Those without result, those which are locked and those which already have an assessment.**

1. Test the functionality of the exercise-specific dashboard (see navigation below) in exam mode with two correction rounds. 
Under: exam -> exercise-groups -> submissions for a specific exercise

- Assess an exam for a first time, then move back and check if the lock was created. 
- Continue the first assessment
- Submit the first assessment
- Return to the dashboard and check if you can open the first assessment
- now assess the submission in the second correction round. Check if the feedback from the previous round is loaded.
- If so, move back to the dashboard to check if the assessment is locked. Then continue, submit and open it once again. 
- Assessments of other tutors should also be visible in the dashboard. The latest result's score is visible in the result column, as it is the one that counts for the grade.

2. Test the functionality in exam mode with one correction round
It should only display the buttons for one correction round, and those buttons should work correctly. 

3. Test the functionality not in exam mode, but with a regular exercise (text & modeling) 
It should only display the buttons for one correction round, and those buttons should work correctly. 

### Screenshots

image

### CODE REVIEW

**Review by simonzilker — CHANGES_REQUESTED** [review]
When assessing through the Instructor dashboard, the buttons do not show up correctly. 
image

**Review by airelawaleria — COMMENTED** [review]
what @simonzilker said

**Review by JohannesWeiss — CHANGES_REQUESTED** [review]
Tested in testing session and can confirm the aforementioned issues.

**Review by marvinchimeka — COMMENTED** [review]
confirm what @simonzilker said.

**Comment by nicolasruscher** [comment]
> When assessing through the Instructor dashboard, the buttons do not show up correctly.
> image

I updated the handling, it should work fine now! :)

**Review by sebastiantpeter — APPROVED** [review]
Tested on ts3. The buttons now look a little bit different than in the screenshots, but everything seems to work fine.

instructor_dashboard

**Review by airelawaleria — CHANGES_REQUESTED** [review]
Tested in the testing session. For text exercises the instructor dashboard correctly displays each assessment , but for the modeling exercises it is not working

P.S. Ts1, Valeryia's Course, TwoCorrectionsAssessment, Modelling exercises
image

**Review by simonzilker — CHANGES_REQUESTED** [review]
Same problem as @airelawaleria

**Review by marvinchimeka — CHANGES_REQUESTED** [review]
Tested in testing session. confirm what @airelawaleria said.

**Review by JohannesWeiss — CHANGES_REQUESTED** [review]
Tested in testing session, can confirm the issue mentioned by @airelawaleria .

**Comment by krusche** [comment]
please update the screenshots in the PR description

**Comment by SimonEntholzer** [comment]
> Tested in the testing session. For text exercises the instructor dashboard correctly displays each assessment , but for the modeling exercises it is not working
> 
> P.S. Ts1, Valeryia's Course, TwoCorrectionsAssessment, Modelling exercises
> image

we fixed the issue, now everything should work as expected :)

**Review by airelawaleria — APPROVED** [review]
tested again on ts1. Issues resolved

**Review by simonzilker — DISMISSED** [review]
Did not find any issues :)

File: src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java [hunk 1]
```diff
@@ -228,18 +228,14 @@ public TextSubmission save(TextSubmission textSubmission) {
      * @return a list of text submissions for the given exercise id
      */
     public List<TextSubmission> getTextSubmissionsByExerciseId(Long exerciseId, boolean submittedOnly, boolean examMode, int correctionRound) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:230]: It seems that the parameter `correctionRound` is not used any more, so we could remove it.

**Review by krusche — CHANGES_REQUESTED** [review]
Code looks good (did not test it yet), I only found a method `getTextSubmissionsByExerciseId` in `TextSubmissionService.java` in which the parameter `correctionRound` can be removed as it is not used any more.

Please also add to test Testing steps that testers should double check that all text/modeling submissions are shown even if they do not have a result yet

File: src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java (hunk 1 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java]: You are right, it's not used anymore. I removed it.

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.html [hunk 2]
```diff
@@ -167,86 +170,128 @@ <h5 *ngIf="modelingExercise.assessmentType === AssessmentType.SEMI_AUTOMATIC &&
...
-                            <button
-                                *ngIf="submission && submission.latestResult && !submission.latestResult!.completionDate"
-                                (click)="cancelAssessment(submission)"
-                                [disabled]="busy"
-                                class="btn btn-outline-secondary btn-sm mb-1"
-                            >
-                                <fa-icon [icon]="'ban'" [fixedWidth]="true"></fa-icon>&nbsp;{{ 'artemisApp.assessment.dashboard.actions.cancel' | translate }}
-                            </button>
+                            <!-- exercise without exam-->
+                            <span *ngIf="modelingExercise.type === ExerciseType.MODELING && submission && modelingExercise.course">
```
Comment (ivanchimeno) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.html:177]: That if statement is big and used in a couple of places. You can technically extract that into an `isExerciseWithoutExam()` function

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts [hunk 3]
```diff
@@ -90,6 +91,8 @@ export class ModelingAssessmentDashboardComponent implements OnInit, OnDestroy {
                 if (res.body!.type === ExerciseType.MODELING) {
                     this.modelingExercise = res.body as ModelingExercise;
                     this.getSubmissions(true);
+                    this.numberOfCorrectionrounds = this.modelingExercise.exerciseGroup ? this.modelingExercise!.exerciseGroup.exam!.numberOfCorrectionRoundsInExam! : 1;
+                    this.setPermissions();
                 } else {
                     // TODO: error message if this is not a modeling exercise
```
Comment (ivanchimeno) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts:97]: Did you forget to implement this or is this intended? :D

Comment (nicolasruscher) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts:97]: This statement was there before us, but yes we could remove it and implement a message

**Review by ivanchimeno — CHANGES_REQUESTED** [review]
Code looks good, just added some minor comments to avoid code duplication.

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts (hunk 3 above)
Comment (ivanchimeno) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts:97]: Ah no I thought it was from you

Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.ts:97]: That was there before :D I'll check how this is handled in other exercise types, and then I'll either delete the else or add an error.

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.html [hunk 4]
```diff
@@ -167,86 +170,128 @@ <h5 *ngIf="modelingExercise.assessmentType === AssessmentType.SEMI_AUTOMATIC &&
...
+                                <button
+                                    *ngIf="!submission.latestResult"
+                                    [disabled]="busy"
+                                    [queryParams]="{ optimal: false, hideBackButton: true }"
+                                    [routerLink]="[
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.html]: I created a method to make this more compact `getAssessmentLink(submission.id)` :)

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.html (hunk 2 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-dashboard.component.html:177]: I don't think that's necessary here, as it does not simplify the html a lot.

**Review by ivanchimeno — APPROVED** [review]

**Review by simonzilker — APPROVED** [review]
re-approve

**Review by LudwigWeide — APPROVED** [review]

**Review by marvinchimeka — APPROVED** [review]

**Review by EneaGore — APPROVED** [review]

**Review by airelawaleria — APPROVED** [review]

**Review by JohannesWeiss — APPROVED** [review]
Re-Approve

**Review by krusche — APPROVED** [review]

**Pull request merged**